src/sleep/sleep.c

```diff
@@ -263,40 +263,66 @@ static int execute(
-                  FORMAT_TIMESPAN(sleep_config->hibernate_delay_sec, USEC_PER_SEC));
+                tfd = timerfd_create(CLOCK_BOOTTIME_ALARM, TFD_NONBLOCK|TFD_CLOEXEC);
+                if (tfd < 0)
+                        return log_error_errno(errno, "Error creating timerfd: %m");
 
-        timespec_store(&ts.it_value, sleep_config->hibernate_delay_sec);
+                /* Calculate current battery capacity and current time before suspension*/
+                last_capacity = read_battery_capacity();
+                before_timestamp = now(CLOCK_MONOTONIC);
 
-        r = timerfd_settime(tfd, 0, &ts, NULL);
-        if (r < 0)
-                return log_error_errno(errno, "Error setting hibernate timer: %m");
+                log_debug("Set timerfd wake battery estimate alarm for %s",
+                        FORMAT_TIMESPAN(suspend_interval, USEC_PER_SEC));
 
-        r = execute(sleep_config, SLEEP_SUSPEND, NULL);
-        if (r < 0)
-                return r;
+                timespec_store(&ts.it_value, suspend_interval);
 
-        r = fd_wait_for_event(tfd, POLLIN, 0);
-        if (r < 0)
-                return log_error_errno(r, "Error polling timerfd: %m");
-        if (!FLAGS_SET(r, POLLIN)) /* We woke up before the alarm time, we are done. */
-                return 0;
+                r = timerfd_settime(tfd, 0, &ts, NULL);
+                if (r < 0)
+                        return log_error_errno(errno, "Error setting battery estimate timer: %m");
+
+                r = execute(sleep_config, SLEEP_SUSPEND, NULL);
+                if (r < 0)
+                        return r;
+
+                r = fd_wait_for_event(tfd, POLLIN, 0);
+                if (r < 0)
+                        return log_error_errno(r, "Error polling timerfd: %m");
+                if (!FLAGS_SET(r, POLLIN))  { /* We woke up before the alarm time, calculate battery discharged and estimate rate. */
+                        after_timestamp = now(CLOCK_MONOTONIC);
+                        current_capacity = read_battery_capacity();
+                        log_debug("%d percentage battery discharged in %s time",
+                                        last_capacity - current_capacity,
+                                        FORMAT_TIMESPAN(after_timestamp - before_timestamp, 0));
+                        estimated_total_discharge_time = (current_capacity*(before_timestamp-after_timestamp))/(last_capacity-current_capacity);
+                        log_debug("Battery will discharge in %d time", estimated_total_discharge_time);
+                        return 0;
+                }
+                tfd = safe_close(tfd);
+
+                /* If woken up after alarm time, calculate battery capacity and battery capacity_level and estimate discharge rate */
+
+                current_capacity = read_battery_capacity();
+
+                estimated_total_discharge_time = current_capacity/(last_capacity-current_capacity);
```

Comment:
`suspend_interval` needs to be updated with the result of this calculation - at the moment it's just used for the log message

Reply:
hmm. yes. i added L320 for test purpose. commented below.